cabal-install/Distribution/Client/ProjectBuilding.hs

```diff
@@ -1041,29 +1043,34 @@ buildInplaceUnpackedPackage verbosity
           annotateFailureNoLog BuildFailed $
             setup buildCommand buildFlags buildArgs
 
-          --TODO: [required eventually] this doesn't track file
-          --non-existence, so we could fail to rebuild if someone
-          --adds a new file which changes behavior.
-          allSrcFiles <-
-            let trySdist    = allPackageSourceFiles verbosity scriptOptions srcdir
-                -- This is just a hack, to get semi-reasonable file
-                -- listings for the monitor
-                tryFallback = do
-                    warn verbosity $
-                        "Couldn't use sdist to compute source files; falling " ++
-                        "back on recursive file scan."
-                    filter (not . ("dist" `isPrefixOf`))
-                        `fmap` getDirectoryContentsRecursive srcdir
-            in if elabSetupScriptCliVersion pkg >= mkVersion [1,17]
-                  then do r <- trySdist
-                          if null r
-                            then tryFallback
-                            else return r
-                  else tryFallback
+          monitors <- case PD.buildType (elabPkgDescription pkg) of
+            -- It's simple! Use SourceFiles
+            Just Simple ->
+              -- TODO: Would be nice if we could have used
+              -- rerunIfChanged directly...
+              execRebuild srcdir (needElaboratedConfiguredPackage pkg)
+            -- Otherwise, operate conservatively
+            _ ->
+                let trySdist    = allPackageSourceFiles verbosity scriptOptions srcdir
+                    -- This is just a hack, to get semi-reasonable file
+                    -- listings for the monitor
+                    tryFallback = do
```

Comment:
Our fallback could now be the `needElaboratedConfiguredPackage` rather than the really dumb `getDirectoryContentsRecursive`